File: SICP/Lecture_6_full/Code/Python/delaystream/requeens.py
## Eight Queens Problem, Recursive, list stream
import copy

from stream import *

def solveall(n):
    solutions = collect_stream(fillrows(n, n))
    return solutions if solutions != [] else None

def fillrows(size, row):
    if row == 0:
        return cons_stream(emptyboard(), lambda : the_empty_stream())

    return filter_stream(lambda poses: issafe(poses), 
                         flatmap(lambda restposes: map_stream(lambda col: adjoinpos(row, col, restposes), 
                                                              enum_interval(1, size)), 
                                         fillrows(size, row - 1)))

def solveall2(n):
    solutions = collect_stream(fillrows2(n, n))
    return solutions if solutions != [] else None

def fillrows2(size, row):
    if row == 0:
        return cons_stream(emptyboard(), lambda : the_empty_stream())

    return filter_stream(lambda poses: issafe(poses), 
                         flatmap(lambda col: map_stream(lambda restposes: adjoinpos(row, col, restposes), 
                                                        fillrows2(size, row - 1)), 
                                 enum_interval(1, size)))

def emptyboard():
    return [];

def adjoinpos(currow, curcol, curposes):
    newposes = copy.copy(curposes)
    newposes.append((currow, curcol))
    return newposes
    
def issafe(poses):
    newpos = poses[-1]
    restposes = poses[:-1]
    return is_not_samecol(newpos, restposes) and is_not_diagonal(newpos, restposes)

def is_not_samecol(newpos, restposes):
    return all(map(lambda restpos: newpos[1] != restpos[1], restposes))

def is_not_diagonal(newpos, restposes):
    return all(map(lambda restpos: abs(newpos[0] - restpos[0]) != abs(newpos[1] - restpos[1]), restposes))

## Tests
def test_solveall():
    assert [[(1, 2), (2, 4), (3, 1), (4, 3)],
            [(1, 3), (2, 1), (3, 4), (4, 2)]] == solveall(4)
    assert None == solveall(3)
    assert 92 == len(solveall(8))

    print 'test_solveall ok'

def test_solveall2():
    assert [[(1, 3), (2, 1), (3, 4), (4, 2)],
            [(1, 2), (2, 4), (3, 1), (4, 3)]] == solveall2(4)
    assert None == solveall2(3)
    # assert 92 == len(solveall2(8))

    print 'test_solveall2 ok'    

def test():
    test_solveall()
    test_solveall2()

    print 'test ok'

if __name__ == '__main__':
    test()

File: SICP/Lecture_6_full/Code/Java/Stream/src/fayelab/sicp/stream/delaystreamclass/ReQueens.java
package fayelab.sicp.stream.delaystreamclass;

import java.util.ArrayList;
import java.util.List;
import java.util.Optional;

import static java.util.Arrays.asList;
import static fayelab.sicp.stream.delaystreamclass.StreamOp.*;

public class ReQueens
{
    public static Optional<List<List<List<Integer>>>> solveAll(int n)
    {
        List<List<List<Integer>>> solutions = collectStream(fillRows(n, n));
        return solutions.isEmpty() ? Optional.empty() : Optional.of(solutions);
    }
    
    private static Stream<List<List<Integer>>> fillRows(int size, int row)
    {
        if(row == 0)
        {
            return consStream(emptyBoard(), () -> theEmptyStream());
        }
        
        return filterStream((List<List<Integer>> poses) -> isSafe(poses), 
                            flatmap((List<List<Integer>> restPoses) -> mapStream((Integer col) -> adjoinPos(row, col, restPoses), 
                                                                                 enumInterval(1, size)), 
                                    fillRows(size, row - 1)));
    }
    
    public static Optional<List<List<List<Integer>>>> solveAll2(int n)
    {
        List<List<List<Integer>>> solutions = collectStream(fillRows2(n, n));
        return solutions.isEmpty() ? Optional.empty() : Optional.of(solutions);
    }
    
    private static Stream<List<List<Integer>>> fillRows2(int size, int row)
    {
        if(row == 0)
        {
            return consStream(emptyBoard(), () -> theEmptyStream());
        }
        
        return filterStream((List<List<Integer>> poses) -> isSafe(poses), 
                            flatmap((Integer col) -> mapStream((List<List<Integer>> restPoses) -> adjoinPos(row, col, restPoses), 
                                                                fillRows2(size, row - 1)),
                                    enumInterval(1, size)));
    }

    private static List<List<Integer>> emptyBoard()
    {
        return new ArrayList<>();
    }

    private static List<List<Integer>> adjoinPos(int curRow, int curCol, List<List<Integer>> curPoses)
    {
        List<List<Integer>> result = new ArrayList<>();
        result.addAll(curPoses);
        result.add(asList(curRow, curCol));
        return result;
    }
    
    private static boolean isSafe(List<List<Integer>> poses)
    {
        List<Integer> newPos = poses.get(poses.size() - 1);
        List<List<Integer>> restPoses = poses.subList(0, poses.size() - 1);
        return isNotSameCol(newPos, restPoses) && isNotDiagonal(newPos, restPoses);
    }

    private static boolean isNotSameCol(List<Integer> newPos, List<List<Integer>> restPoses)
    {
        return restPoses.stream()
                        .noneMatch(restPos -> newPos.get(1) == restPos.get(1));
    }

    private static boolean isNotDiagonal(List<Integer> newPos, List<List<Integer>> restPoses)
    {
        return restPoses.stream()
                        .noneMatch(restPos -> Math.abs(newPos.get(0) - restPos.get(0)) 
                                                == Math.abs(newPos.get(1) - restPos.get(1)));
    }
}
